docs/tutorials/tpu-cluster-setup.md

```diff
@@ -0,0 +1,135 @@
+
+## TPU Cluster Architecture
+
+A typical TPU cluster consists of:
+- **Head Node**: A persistent GCP VM that coordinates the cluster
+- **Worker Nodes**: Autoscaling TPU VMs that execute the actual computation
+
+## Cluster Configuration
+
+The cluster configuration is defined in a [YAML template file](../../infra/marin-cluster-template.yaml). This template allows you to customize various aspects of your cluster:
+
+- Machine types for head and worker nodes
+- Number of CPUs and amount of RAM
+- Disk size and type
+- TPU types and versions
+- Minimum and maximum worker counts
+- Docker images and initialization commands
+- Network configuration
+- Environment variables and secrets
+
+For our clusters, each config is in a separate file in the `infra` directory. These files are automatically generated by the
+`infra/update-cluster-configs.py` script, which reads the `infra/marin-cluster-template.yaml` file.
+More information can be found in the [infra/README.md](../../infra/README.md).
+
+## Setting Up the Cluster
+
+1. Install Marin with TPU support:
+   ```bash
+   pip install -e ".[tpu]"
+   ```
+
+2. Set up your GCP environment:
+   ```bash
+   # Configure Docker for GCP artifact registry
+   gcloud auth configure-docker us-central2-docker.pkg.dev
+   gcloud auth configure-docker europe-west4-docker.pkg.dev
+   gcloud auth configure-docker us-west4-docker.pkg.dev
+   ```
+
+3. Start the cluster (replace `$CLUSTER` with your desired region, e.g., `us-central2`):
+   ```bash
+   ray up -y infra/marin-$CLUSTER.yaml
+   ```
+
+## Executing a Job on the cluster
+
+### Using Ray Submit
+
+1. Connect to the cluster using ray dashboard:
+   ```bash
+   ray dashboard infra/marin-$CLUSTER.yaml
+   ```
+2. In another terminal, submit a job using ray submit:
+   ```bash
+   ray job submit --working-dir . -- python experiments/tutorials/hello_world.py
+   ```
+
+### Using Ray Run
+
+1. Connect to the cluster using ray dashboard:
+   ```bash
+   ray dashboard infra/marin-$CLUSTER.yaml
+   ```
+2. We have made a thin wrapper on top of `ray submit` called [ray-run](../../marin/run/ray_run.py) which can be used to easily specify the pip requirements and environment variables.
+   ```bash
+   python marin/run/ray_run.py --env_vars WANDB_API_KEY ${WANDB_API_KEY}  --pip_deps jax==0.4.35,async-lru --  python experiments/tutorials/hello_world.py
+   ```
+
+## Managing the Cluster
+
+### Basic Commands
+
+```bash
+# Monitor cluster status
+ray status infra/marin-$CLUSTER.yaml
+
+# SSH into head node
+ray attach infra/marin-$CLUSTER.yaml
+
+# View autoscaler logs
+ray exec infra/marin-$CLUSTER.yaml "tail -n 100 -f /tmp/ray/session_latest/logs/monitor*"
+
+# Stop the cluster
+ray down -y infra/marin-$CLUSTER.yaml
+```
+
+## Best Practices
+
+1. **Data Location**: Use the bucket in the same region as your cluster to minimize data transfer costs.
+
+2. **Preemption Handling**: Most compute is preemptible, meaning VMs can be shut down at any time. Design your jobs to:
```

Comment:
people might not have preemptible compute, so say "if your compute is preemptible"